src/librustc_codegen_llvm/coverageinfo/mod.rs

```diff
@@ -100,107 +105,150 @@ impl CoverageInfoBuilderMethods<'tcx> for Builder<'a, 'll, 'tcx> {
         let mut coverage_regions = self.coverage_context().function_coverage_map.borrow_mut();
         coverage_regions
             .entry(instance)
-            .or_insert_with(|| {
-                FunctionCoverage::with_coverageinfo(self.tcx.coverageinfo(instance.def_id()))
-            })
-            .add_unreachable(start_byte_pos, end_byte_pos);
+            .or_insert_with(|| FunctionCoverage::new(self.tcx, instance))
+            .add_unreachable_region(start_byte_pos, end_byte_pos);
     }
 }
 
-/// This struct wraps an opaque reference to the C++ template instantiation of
-/// `llvm::SmallVector<coverage::CounterExpression>`. Each `coverage::CounterExpression` object is
-/// constructed from primative-typed arguments, and pushed to the `SmallVector`, in the C++
-/// implementation of `LLVMRustCoverageSmallVectorCounterExpressionAdd()` (see
-/// `src/rustllvm/CoverageMappingWrapper.cpp`).
-pub struct SmallVectorCounterExpression<'a> {
-    pub raw: &'a mut llvm::coverageinfo::SmallVectorCounterExpression<'a>,
-}
+/// Aligns to C++ struct llvm::coverage::Counter::CounterKind.
+/// The order of discrimiators is important.
+#[derive(Copy, Clone, Debug)]
+#[repr(C)]
+enum RegionKind {
+    /// A CodeRegion associates some code with a counter
+    CodeRegion,
```

Comment:
```suggestion
    CodeRegion = 0,
```

etc

Reply:
@wesleywiser Will do.

Also, what do you think about providing a link to the in-tree LLVM source this is mapping to? For example:

```rust
/// Aligns to C++ struct [llvm::coverage::Counter::CounterKind](https://github.com/rust-lang/llvm-project/blob/rustc/10.0-2020-05-05/llvm/include/llvm/ProfileData/Coverage/CoverageMapping.h#L203-L221)
```

Should result in rustdoc:

> Aligns to C++ struct [llvm::coverage::Counter::CounterKind](https://github.com/rust-lang/llvm-project/blob/rustc/10.0-2020-05-05/llvm/include/llvm/ProfileData/Coverage/CoverageMapping.h#L203-L221)

I think it's particularly convenient to see exactly what the originating C++ definition was.

Originally, I left it out because there is a slight risk someone might assume the link maps to the _current_ Rust snapshot, which could be misleading if, for some reason, one of the Rust-replicated enum or struct definitions diverges, in some meaningful way, in a future version/shapshot.

Rust's fork of LLVM doesn't use `master` apparently (per https://github.com/rust-lang/llvm-project/blob/master/README.md) and there's no `latest` tag or anything like that, from what I see. So I don't think there's a way to make this a version-less link.

Reply:
(Note: I made some small edits to the comment above, fixing errors in the link text and URL start line)

Reply:
I think linking to a specific version is fine. IIRC we do similar things in other places so there's precedent for that.

Reply:
done